src/win/fs.c

```diff
@@ -1611,26 +1613,356 @@ static void fs__access(uv_fs_t* req) {
+      goto accesscheck_cleanup;
+    }
+
+    /*
+     * Next, construct a mapping from generic access rights to
+     * the more specific access rights that AccessCheck expects.
+     */
+    DWORD executeAccessRights = FILE_GENERIC_EXECUTE;
+    GENERIC_MAPPING mapping = { 0xFFFFFFFF };
+    mapping.GenericExecute = FILE_GENERIC_EXECUTE;
+    MapGenericMask(&executeAccessRights, &mapping);
+
+    PRIVILEGE_SET privileges = { 0 };
+    DWORD grantedAccess = 0, privilegesLen = sizeof(privileges);
+    BOOL result = FALSE;
+    if (AccessCheck(sd,
+                    hImpersonatedToken,
+                    executeAccessRights,
+                    &mapping,
+                    &privileges,
+                    &privilegesLen,
+                    &grantedAccess,
+                    &result)) {
+      /*
+       * If AccessCheck passes, nothing went wrong, but
+       * we must still check that we have access.
+       */
+      if (!result) {
+        SET_REQ_WIN32_ERROR(req, UV_EPERM);
+        goto accesscheck_cleanup;
+      }
+    } else {
+      /*
+       * This signifies that something went wrong with the
+       * actual AccessCheck() invocation itself.
+       */
+      SET_REQ_WIN32_ERROR(req, GetLastError());
+      goto accesscheck_cleanup;
+    }
+
+accesscheck_cleanup:
+    uv__free(sd);
+    if (hImpersonatedToken != NULL) {
+      CloseHandle(hImpersonatedToken);
+    }
+    if (hToken != NULL) {
+      CloseHandle(hToken);
+    }
+    /*
+     * If the result is false, return immediately.
+     * Some error code has been set in the `req` already.
+     */
+    if (!result) {
+      return;
+    }
+  }
+
+  /* If we get to the end, everything worked out. */
+  SET_REQ_SUCCESS(req);
 }
 
+static void build_explicit_access(EXPLICIT_ACCESS_W* ea, PSID owner, TRUSTEE_TYPE user_type, mode_t mode_triplet) {
+  /*
+   * We map the typical POSIX mode bits r/w/x as the following windows permissions:
+   * - Read: FILE_READ_DATA | FILE_READ_EA | FILE_READ_ATTRIBUTES
+   * - Write: FILE_WRITE_DATA | FILE_WRITE_EA
+   * - Execute: FILE_EXECUTE | FILE_READ_ATTRIBUTES
+   *
+   * We map the "User", "Group", and "Other" sets as the current object owner,
+   * the current object group, and the "Everyone" group, respectively.
+   */
+  ZeroMemory(ea, 2*sizeof(EXPLICIT_ACCESS_W));
+
+  /* 
+   * Initialize two EXLPICIT_ACCESS structures; one to explicitly allow things, the
+   * other to explicitly deny them.  We leave no middle ground for inheritance to mess
+   * things up.
+   */
+  for(int i=0; i<2; ++i) {
+    ea[i].grfAccessPermissions = 0;
+    ea[i].grfInheritance = NO_INHERITANCE;
+    ea[i].Trustee.MultipleTrusteeOperation = NO_MULTIPLE_TRUSTEE;
+    ea[i].Trustee.TrusteeForm = TRUSTEE_IS_SID;
+    ea[i].Trustee.TrusteeType = user_type;
+    ea[i].Trustee.ptstrName = owner;
+  }
+
+  ea[0].grfAccessMode = GRANT_ACCESS;
+  ea[1].grfAccessMode = DENY_ACCESS;
+  
+  if( mode_triplet & 0x1 ) {
```

Comment:
```suggestion
  if (mode_triplet & 0x1)
```
libuv style guide https://github.com/libuv/libuv/blob/v1.x/CONTRIBUTING.md follows https://google.github.io/styleguide/cppguide.html, but often limits unrequired braces